src/test/java/org/apache/mesos/offer/constrain/AndRuleTest.java

```diff
@@ -0,0 +1,207 @@
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3, RESOURCE_4), o);
+
+        o = new AndRule(REMOVE_FIRST, REMOVE_FIRST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3, RESOURCE_4), o);
+
+        o = new AndRule(REMOVE_FIRST, ALL, REMOVE_FIRST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3, RESOURCE_4), o);
+
+        o = new AndRule(ALL, REMOVE_FIRST, ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3, RESOURCE_4), o);
+
+        o = new AndRule(REMOVE_FIRST, REMOVE_FIRST, REMOVE_FIRST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3, RESOURCE_4), o);
+    }
+
+    @Test
+    public void testLastRemoved() {
+        Offer o = new AndRule(REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_LAST, ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(ALL, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_LAST, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_LAST, ALL, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(ALL, REMOVE_LAST, ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_LAST, REMOVE_LAST, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_1, RESOURCE_2, RESOURCE_3), o);
+    }
+
+    @Test
+    public void testFirstLastRemoved() {
+        Offer o = new AndRule(REMOVE_LAST, REMOVE_FIRST, ALL).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_FIRST, ALL, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_LAST, REMOVE_FIRST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3), o);
+
+        o = new AndRule(REMOVE_FIRST, REMOVE_LAST, REMOVE_FIRST, REMOVE_LAST).filter(offerWith(RESOURCES));
+        assertEquals(offerWith(RESOURCE_2, RESOURCE_3), o);
+    }
+
+    @Test
+    public void testGenerator() {
+        PlacementRule r = new AndRule.Generator(
+                new PassthroughGenerator(REMOVE_LAST),
+                new PassthroughGenerator(REMOVE_FIRST),
+                new PassthroughGenerator(ALL))
+                .generate(Collections.emptyList());
+        Offer o = r.filter(offerWith(RESOURCES));
+        assertEquals(r.toString(), offerWith(RESOURCE_2, RESOURCE_3), o);
```

Comment:
It does get called? @\Test annotation is present..

This is really just a sanity test that the AndRule.Generator utility behaves as expected.